### PR TITLE
Nondimensional formulation

### PR DESCRIPTION
**Pull request by TheBB**
I removed the previous scaling feature, which is why so many test files changed. The only difference is that their pressures are different by a constant factor.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
ignore the failure for non-gcc configuration, it's just that i removed hdf5 from clang configuration so we have a build without it (and we have issues without it right now).

**Comment by TheBB** [comment]
jenkins build this with downstreams please

File: PoroElastic/PoroElasticity.h [hunk 1]
```diff
@@ -103,11 +103,18 @@ class PoroElasticity : public Elasticity
...
          bool residual, int nbasis, int nsd);
+
     //! \brief Empty destructor.
     virtual ~Mats() {}
 
     //! \brief Updates the time step size.
     void setStepSize(double dt) { h = dt; }
+
+    //! \brief Enables nondimensionalization.
+    void nondimensionalize(Characteristics cars) {
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: use surrounding bracing style

File: PoroElastic/PoroElasticity.h [hunk 2]
```diff
@@ -117,10 +124,15 @@ class PoroElasticity : public Elasticity
     virtual const Vector& getRHSVector() const;
 
   protected:
+    virtual void scaleMatrices();
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: doxy

File: PoroElastic/PoroElasticity.h [hunk 3]
```diff
@@ -159,6 +171,14 @@ class PoroElasticity : public Elasticity
     //! \brief Empty destructor.
     virtual ~NewmarkMats() {}
 
+    virtual void scaleMatrices()
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: doxy

File: PoroElastic/PoroElasticity.h [hunk 4]
```diff
@@ -193,24 +216,28 @@ class PoroElasticity : public Elasticity
...
+        P::A[pp_S].multiply(P::vec[Vpvel], tp, -c.p, 1.0);
       if (P::A.size() > up_Q && P::vec.size() > Vuvel)
-        P::A[up_Q].multiply(P::vec[Vuvel], tp, true, -1);
+        P::A[up_Q].multiply(P::vec[Vuvel], tp, - 1 / c.alpha, 1.0, true);
       if (P::A.size() > uu_M && P::vec.size() > Vuvel && m_damp > 0.0)
-        P::A[uu_M].multiply(P::vec[Vuvel]*m_damp, tu, false, -1);
+        P::A[uu_M].multiply(P::vec[Vuvel]*m_damp, tu, -1 / c.E / c.t / c.t, 1.0);
       if (P::A.size() > uu_K && P::vec.size() > Vuvel && s_damp > 0.0)
-        P::A[uu_K].multiply(P::vec[Vuvel]*s_damp, tu, false, -1);
+        P::A[uu_K].multiply(P::vec[Vuvel]*s_damp, tu, -1 / c.E, 1.0);
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: -1.0

File: PoroElastic/PoroElasticity.h [hunk 5]
```diff
@@ -193,24 +216,28 @@ class PoroElasticity : public Elasticity
...
+        P::A[up_D].multiply(P::vec[Vuacc], tp, 1 / c.alpha / c.t, 1.0, true);
       if (P::A.size() > pp_S && P::vec.size() > Vpvel)
-        P::A[pp_S].multiply(P::vec[Vpvel], tp, false,-1);
+        P::A[pp_S].multiply(P::vec[Vpvel], tp, -c.p, 1.0);
       if (P::A.size() > up_Q && P::vec.size() > Vuvel)
-        P::A[up_Q].multiply(P::vec[Vuvel], tp, true, -1);
+        P::A[up_Q].multiply(P::vec[Vuvel], tp, - 1 / c.alpha, 1.0, true);
       if (P::A.size() > uu_M && P::vec.size() > Vuvel && m_damp > 0.0)
-        P::A[uu_M].multiply(P::vec[Vuvel]*m_damp, tu, false, -1);
+        P::A[uu_M].multiply(P::vec[Vuvel]*m_damp, tu, -1 / c.E / c.t / c.t, 1.0);
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: -1.0

File: PoroElastic/PoroElasticity.h [hunk 6]
```diff
@@ -193,24 +216,28 @@ class PoroElasticity : public Elasticity
...
+        P::A[up_Q].multiply(P::vec[Vp],    tu, c.p / c.E, 1.0);
       if (P::A.size() > pp_P && P::vec.size() > Vp)
-        P::A[pp_P].multiply(P::vec[Vp],    tp, false,-1);
+        P::A[pp_P].multiply(P::vec[Vp],    tp, -c.t * c.p / c.alpha, 1.0);
       if (P::A.size() > uu_M && P::vec.size() > Vuacc)
-        P::A[uu_M].multiply(P::vec[Vuacc], tu, false,-1);
+        P::A[uu_M].multiply(P::vec[Vuacc], tu, - 1 / c.E / c.t / c.t, 1.0);
       if (P::A.size() > up_D && P::vec.size() > Vuacc && !P::A[up_D].empty())
-        P::A[up_D].multiply(P::vec[Vuacc], tp, true,  1);
+        P::A[up_D].multiply(P::vec[Vuacc], tp, 1 / c.alpha / c.t, 1.0, true);
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: -1.0

File: PoroElastic/PoroElasticity.h [hunk 7]
```diff
@@ -193,24 +216,28 @@ class PoroElasticity : public Elasticity
...
+
       if (P::A.size() > up_Q && P::vec.size() > Vp)
-        P::A[up_Q].multiply(P::vec[Vp],    tu, false, 1);
+        P::A[up_Q].multiply(P::vec[Vp],    tu, c.p / c.E, 1.0);
       if (P::A.size() > pp_P && P::vec.size() > Vp)
-        P::A[pp_P].multiply(P::vec[Vp],    tp, false,-1);
+        P::A[pp_P].multiply(P::vec[Vp],    tp, -c.t * c.p / c.alpha, 1.0);
       if (P::A.size() > uu_M && P::vec.size() > Vuacc)
-        P::A[uu_M].multiply(P::vec[Vuacc], tu, false,-1);
+        P::A[uu_M].multiply(P::vec[Vuacc], tu, - 1 / c.E / c.t / c.t, 1.0);
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: -1.0/(c.E*c.t*c.t)

File: PoroElastic/PoroElasticity.h [hunk 8]
```diff
@@ -423,6 +442,11 @@ class PoroElasticity : public Elasticity
   bool residual;        //!< If \e true, compute residuals
 
 private:
+  int ndim;             //!< Number of spatial dimensions
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: available as member nsd from IntegrandBase.

File: PoroElastic/PoroElasticity.h [hunk 9]
```diff
@@ -18,6 +18,25 @@
...
+  \brief Characteristic quantities for nondimensionalization.
+*/
+struct Characteristics {
+  double E;         //!< Characteristic Young's modulus
+  double alpha;     //!< Characteristic Biot's coefficient
+  double perm;      //!< Characteristic permeability
+  double p;         //!< Characteristic pressure
+  double t;         //!< Characteristic time
+
+  Characteristics() : E(1.0), alpha(1.0), perm(1.0), p(1.0), t(1.0) {};
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: doxy.

File: PoroElastic/PoroElasticity.h [hunk 10]
```diff
@@ -18,6 +18,25 @@
...
+
+  void normalize() {
```
Comment (akva2) [inline PoroElastic/PoroElasticity.h]: doxy

**Comment by TheBB** [comment]
Thanks, updated.

File: PoroElastic/PoroElasticity.C [hunk 11]
```diff
@@ -137,9 +137,9 @@ bool PoroElasticity::init (const TimeDomain& time)
 
     if (cars.perm == 0.0) {
       Vec3 perm = pmat->getPermeability(X);
-      for (int d = 1; d <= ndim; d++)
+      for (int d = 1; d <= nsd; d++)
```
Comment (akva2) [inline PoroElastic/PoroElasticity.C]: nsd is unsigned, i'm surprised this does not yield comparison warnings.

File: PoroElastic/PoroMats.C [hunk 12]
```diff
@@ -88,16 +107,20 @@ const Vector& PoroElasticity::Mats::getRHSVector () const
...
   // RHS arising from the previous timestep.
   // Fp' = Fp * h + Q_pu * u + S_pp * p
+  // Note that the nondimensionalization of the matrices in getNewtonMatrix()
+  // hasn't happened yet when this method runs, so we replicate those here.
   if (dynamic == 1) {
     Vector& fp = const_cast<Vector&>(b[Fp]);
     fp *= h;
     if (A.size() > up_Q && vec.size() > Vu)
-      A[up_Q].multiply(vec[Vu], fp, true, true);
+      A[up_Q].multiply(vec[Vu], fp, 1 / cars.alpha, 1.0, true);
```
Comment (akva2) [inline PoroElastic/PoroMats.C]: 1.0

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by TheBB** [comment]
I'm confused, there are tests in OpenFrac that will fail with this PR but jenkins doesn't seem to complain about it.

**Comment by akva2** [comment]
yes, jenkins downstream list had not been updated, see #39

**Pull request merged**